### ISSUE TITLE
[BUG]: Error: Unrecognized argument '"-setParam:name='IIS Web Application Name',value='{REDACTED}'"'. All arguments must begin with "-"

### ISSUE THREAD

**Issue by BRGROSU** [microsoft/azure-pipelines-tasks#18817]
### Task name

AzureRmWebAppDeployment

### Task version

4.226.0

### Environment type (Please select at least one enviroment where you face this issue)

- [ ] Self-Hosted
- [ ] Microsoft Hosted
- [X] VMSS Pool
- [ ] Container

### Azure DevOps Server type

dev.azure.com (formerly visualstudio.com)

### Operation system

Windows Server 2019 Datacenter

### Task log

```shell
2023-08-14T08:48:52.9016125Z ##[debug]Constructed msDeploy comamnd line arguments
2023-08-14T08:48:52.9311743Z ##[debug]System.DefaultWorkingDirectory=C:\a\5\s
2023-08-14T08:48:52.9314007Z ##[debug]the argument string is:
2023-08-14T08:48:52.9315440Z ##[debug] -verb:sync -source:package="'C:\a\5\{REDACTED}\{REDACTED}.zip'" -dest:auto,ComputerName="'https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}'",UserName="'${REDACTED}'",Password="'{REDACTED}'",AuthType="'Basic'" -setParam:name="'IIS Web Application Name'",value="'{REDACTED}'" -enableRule:AppOffline -setParamFile=1692002930883_tempSetParameters.xml  -skip:Directory=App_Data -retryAttempts:6 -retryInterval:10000 -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
2023-08-14T08:48:52.9316698Z ##[debug]converting the argument string into an array of arguments
2023-08-14T08:48:52.9317154Z ##[debug]the array of arguments is:
2023-08-14T08:48:52.9317540Z ##[debug]arg#0: -verb:sync
2023-08-14T08:48:52.9317946Z ##[debug]arg#1: -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip'
2023-08-14T08:48:52.9318604Z ##[debug]arg#2: -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'
2023-08-14T08:48:52.9319275Z ##[debug]arg#3: -setParam:name='IIS Web Application Name',value='{REDACTED}'
2023-08-14T08:48:52.9319744Z ##[debug]arg#4: -enableRule:AppOffline
2023-08-14T08:48:52.9320168Z ##[debug]arg#5: -setParamFile=1692002930883_tempSetParameters.xml
2023-08-14T08:48:52.9320594Z ##[debug]arg#6: -skip:Directory=App_Data
2023-08-14T08:48:52.9320982Z ##[debug]arg#7: -retryAttempts:6
2023-08-14T08:48:52.9321389Z ##[debug]arg#8: -retryInterval:10000
2023-08-14T08:48:52.9321851Z ##[debug]arg#9: -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
2023-08-14T08:48:52.9322281Z ##[debug]which 'msdeploy'
2023-08-14T08:48:52.9322781Z ##[debug]found: 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:52.9324057Z ##[debug]C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe arg: ["-verb:sync","-source:package='C:\\a\\5\\{REDACTED}\\{REDACTED}.zip'","-dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'","-setParam:name='IIS Web Application Name',value='{REDACTED}'","-enableRule:AppOffline","-setParamFile=1692002930883_tempSetParameters.xml","-skip:Directory=App_Data","-retryAttempts:6","-retryInterval:10000","-userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0"]
2023-08-14T08:48:52.9326154Z ##[debug]C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe arg: ["-verb:sync","-source:package='C:\\a\\5\\{REDACTED}\\{REDACTED}.zip'","-dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'","-setParam:name='IIS Web Application Name',value='{REDACTED}'","-enableRule:AppOffline","-setParamFile=1692002930883_tempSetParameters.xml","-skip:Directory=App_Data","-retryAttempts:6","-retryInterval:10000","-userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0"]
2023-08-14T08:48:52.9327269Z ##[debug]exec tool: C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe
2023-08-14T08:48:52.9327771Z ##[debug]exec tool: C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe
2023-08-14T08:48:52.9328132Z ##[debug]arguments:
2023-08-14T08:48:52.9328433Z ##[debug]arguments:
2023-08-14T08:48:52.9328754Z ##[debug]   -verb:sync
2023-08-14T08:48:52.9329067Z ##[debug]   -verb:sync
2023-08-14T08:48:52.9329420Z ##[debug]   -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip'
2023-08-14T08:48:52.9329827Z ##[debug]   -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip'
2023-08-14T08:48:52.9330446Z ##[debug]   -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'
2023-08-14T08:48:52.9331193Z ##[debug]   -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'
2023-08-14T08:48:52.9332288Z ##[debug]   -setParam:name='IIS Web Application Name',value='{REDACTED}'
2023-08-14T08:48:52.9332730Z ##[debug]   -setParam:name='IIS Web Application Name',value='{REDACTED}'
2023-08-14T08:48:52.9333124Z ##[debug]   -enableRule:AppOffline
2023-08-14T08:48:52.9333482Z ##[debug]   -enableRule:AppOffline
2023-08-14T08:48:52.9333848Z ##[debug]   -setParamFile=1692002930883_tempSetParameters.xml
2023-08-14T08:48:52.9334234Z ##[debug]   -setParamFile=1692002930883_tempSetParameters.xml
2023-08-14T08:48:52.9334610Z ##[debug]   -skip:Directory=App_Data
2023-08-14T08:48:52.9334965Z ##[debug]   -skip:Directory=App_Data
2023-08-14T08:48:52.9335313Z ##[debug]   -retryAttempts:6
2023-08-14T08:48:52.9335661Z ##[debug]   -retryAttempts:6
2023-08-14T08:48:52.9336003Z ##[debug]   -retryInterval:10000
2023-08-14T08:48:52.9336344Z ##[debug]   -retryInterval:10000
2023-08-14T08:48:52.9336734Z ##[debug]   -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
2023-08-14T08:48:52.9337173Z ##[debug]   -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
2023-08-14T08:48:52.9338217Z [command]"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip' -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic' "-setParam:name='IIS Web Application Name',value='{REDACTED}'" -enableRule:AppOffline -setParamFile=1692002930883_tempSetParameters.xml -skip:Directory=App_Data -retryAttempts:6 -retryInterval:10000 -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
2023-08-14T08:48:53.5341469Z ##[debug]Exit code 4294967295 received from tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:53.5342031Z ##[debug]Exit code 4294967295 received from tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:53.5342515Z ##[debug]STDIO streams have closed for tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:53.5342808Z ##[debug]STDIO streams have closed for tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:53.5346960Z ##[debug]System.DefaultWorkingDirectory=C:\a\5\s
2023-08-14T08:48:53.5350961Z ##[debug]rm -rf 1692002930883_tempSetParameters.xml
2023-08-14T08:48:53.5352445Z ##[debug]removing file 1692002930883_tempSetParameters.xml
2023-08-14T08:48:53.5514076Z ##[debug]Deployment Failed with Error: Error: Error: Unrecognized argument '"-setParam:name='IIS Web Application Name',value='{REDACTED}'"'. All arguments must begin with "-".
Error count: 1.

2023-08-14T08:48:53.5514500Z ##[debug]task result: Failed
2023-08-14T08:48:53.5541343Z ##[error]Error: Error: Unrecognized argument '"-setParam:name='IIS Web Application Name',value='{REDACTED}'"'. All arguments must begin with "-".
Error count: 1.

2023-08-14T08:48:53.5546977Z ##[debug]Processed: ##vso[task.issue type=error;]Error: Error: Unrecognized argument '"-setParam:name='IIS Web Application Name',value='{REDACTED}'"'. All arguments must begin with "-".
Error count: 1.

2023-08-14T08:48:53.5548210Z ##[debug]Processed: ##vso[task.complete result=Failed;]Error: Error: Unrecognized argument '"-setParam:name='IIS Web Application Name',value='{REDACTED}'"'. All arguments must begin with "-".
Error count: 1.
```

### Relevant log output

```shell
Error: Error: Unrecognized argument '"-setParam:name='IIS Web Application Name',value='{REDACTED}'"'. All arguments must begin with "-"
```

### Aditional info

```shell
4.225.1 version worked as expected. Started failing with 4.226.0.
```

**Comment by BRGROSU** [comment]
Adding these lines to show NodeJS version running on agent and agent versions:
```
##[debug]Using node path: C:\agent\externals\node16\bin\node.exe
```
```
Current agent version: '3.225.0'
```

**Comment by davemandy** [comment]
Until fixed, we are mitigating by pegging the affected pipeline task to an unaffected version, e.g.
`IISWebAppDeploymentOnMachineGroup@0.222.0`
Instead of
`IISWebAppDeploymentOnMachineGroup@0`

**Comment by v-mohithgc** [comment]
Hi @BRGROSU can you please confirm the task in which you're seeing this error? is it AzureRmWebAppDeployment or IISWebAppDeployementOnMachinGrpup?

**Comment by BRGROSU** [comment]
@v-mohithgc AzureRmWebAppDeployment. Until today, our agents used 4.225.1 and they started upgrading to 4.226.0 when the error started. We were using the latest @4 and had to downgrade all our tasks to 4.225.1 manually to not be blocked by it. I believe this is the Node16 task they started upgrading to https://github.com/microsoft/azure-pipelines-tasks/tree/master/_generated/AzureRmWebAppDeploymentV4_Node16

**Comment by akozak-mirango** [comment]
I am also experiencing the exact same issue. We use DevOps pipelines and I have tried using version 3.* of the Azure App Service deploy task, but I had the same issue with 3.226.0.

**Comment by DmitriiBobreshev** [comment]
Hi @BRGROSU, and @akozak-mirango, thank you for the feedback, we're preparing a fix for the tasks, and I'll keep you updated

**Comment by BRGROSU** [comment]
I believe the issue lies where it executes the `msdeploy` cmd, in the ToolRunner. When it constructs the argument string in the `AzureRmWebAppDeployment` task, the parameter seems to be set correctly, without the `""` when the flow exists `getMSDeployCmdArgs` [here](https://github.com/microsoft/azure-pipelines-tasks-common-packages/blob/06c6234c71285be0fe9f3b544d1245aecaf5de47/common-npm-packages/webdeployment-common/msdeployutility.ts#L29) as in `executeMSDeploy` [here](https://github.com/microsoft/azure-pipelines-tasks-common-packages/blob/06c6234c71285be0fe9f3b544d1245aecaf5de47/common-npm-packages/webdeployment-common/deployusingmsdeploy.ts#L154C16-L154C31), the debug argument string is presented as expected like the one below.
 ```
2023-08-14T08:48:51.6762058Z ##[debug]the argument string is:
2023-08-14T08:48:51.6763575Z ##[debug] -verb:sync -source:package="'C:\a\5\{REDACTED}\{REDACTED}.zip'" -dest:auto,ComputerName="'https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}'",UserName="'${REDACTED}'",Password="'{REDACTED}'",AuthType="'Basic'" -setParam:name="'IIS Web Application Name'",value="'{REDACTED}'" -enableRule:AppOffline -setParamFile=1692002930883_tempSetParameters.xml  -skip:Directory=App_Data -retryAttempts:6 -retryInterval:10000 -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
```
After it passes the `converting the argument string into an array of arguments` line, the `argStringToArray` [here](https://github.com/microsoft/azure-pipelines-tasks-common-packages/blob/06c6234c71285be0fe9f3b544d1245aecaf5de47/common-npm-packages/webdeployment-common/deployusingmsdeploy.ts#L172C27-L172C43) seems to be working as expected as well, or else it would have wrapped it with `""` in the output.
```
2023-08-14T08:48:51.6765473Z ##[debug]the array of arguments is:
2023-08-14T08:48:51.6765948Z ##[debug]arg#0: -verb:sync
2023-08-14T08:48:51.6766398Z ##[debug]arg#1: -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip'
2023-08-14T08:48:51.6767135Z ##[debug]arg#2: -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'
2023-08-14T08:48:51.6767838Z ##[debug]arg#3: -setParam:name='IIS Web Application Name',value='{REDACTED}'
2023-08-14T08:48:51.6768270Z ##[debug]arg#4: -enableRule:AppOffline
2023-08-14T08:48:51.6768910Z ##[debug]arg#5: -setParamFile=1692002930883_tempSetParameters.xml
2023-08-14T08:48:51.6769342Z ##[debug]arg#6: -skip:Directory=App_Data
2023-08-14T08:48:51.6769746Z ##[debug]arg#7: -retryAttempts:6
2023-08-14T08:48:51.6775326Z ##[debug]arg#8: -retryInterval:10000
2023-08-14T08:48:51.6775802Z ##[debug]arg#9: -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
```
The next bit seems to be the ToolRunner, where [tl.exec](https://github.com/microsoft/azure-pipelines-tasks-common-packages/blob/06c6234c71285be0fe9f3b544d1245aecaf5de47/common-npm-packages/webdeployment-common/deployusingmsdeploy.ts#L178C11-L178C11). Going deeper into `azure-pipelines-task-lib`, the `task.ts` function that is called seems to be this [exec](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/task.ts#L1450C36-L1450C40) if I am not mistaken. I see the output of `this.tool` being invoked and finding the executable:
```
2023-08-14T08:48:51.6776195Z ##[debug]which 'msdeploy'
2023-08-14T08:48:51.6779604Z ##[debug]found: 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
```
But here it gets interesting. The output is duplicated, not sure if it's because of `tr.on('debug', (data: string) => { debug(data); });`
The ToolRunner then parses the args via `tr.arg` and outputs them [here](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/toolrunner.ts#L812) as an array as it is converted above. But still, the args seem to be ok.

```
2023-08-14T08:48:51.6781784Z ##[debug]C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe arg: ["-verb:sync","-source:package='C:\\a\\5\\{REDACTED}\\{REDACTED}.zip'","-dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'","-setParam:name='IIS Web Application Name',value='{REDACTED}'","-enableRule:AppOffline","-setParamFile=1692002930883_tempSetParameters.xml","-skip:Directory=App_Data","-retryAttempts:6","-retryInterval:10000","-userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0"]
2023-08-14T08:48:51.6783258Z ##[debug]C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe arg: ["-verb:sync","-source:package='C:\\a\\5\\{REDACTED}\\{REDACTED}.zip'","-dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'","-setParam:name='IIS Web Application Name',value='{REDACTED}'","-enableRule:AppOffline","-setParamFile=1692002930883_tempSetParameters.xml","-skip:Directory=App_Data","-retryAttempts:6","-retryInterval:10000","-userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0"]
2023-08-14T08:48:51.6784159Z ##[debug]exec tool: C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe
2023-08-14T08:48:51.6784639Z ##[debug]exec tool: C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe
2023-08-14T08:48:51.6785046Z ##[debug]arguments:
2023-08-14T08:48:51.6785596Z ##[debug]arguments:
2023-08-14T08:48:51.6785949Z ##[debug]   -verb:sync
2023-08-14T08:48:51.6786327Z ##[debug]   -verb:sync
2023-08-14T08:48:51.6786713Z ##[debug]   -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip'
2023-08-14T08:48:51.6787149Z ##[debug]   -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip'
2023-08-14T08:48:51.6787849Z ##[debug]   -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'
2023-08-14T08:48:51.6788648Z ##[debug]   -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic'
2023-08-14T08:48:51.6789126Z ##[debug]   -setParam:name='IIS Web Application Name',value='{REDACTED}'
2023-08-14T08:48:51.6789397Z ##[debug]   -setParam:name='IIS Web Application Name',value='{REDACTED}'
2023-08-14T08:48:51.6789642Z ##[debug]   -enableRule:AppOffline
2023-08-14T08:48:51.6789850Z ##[debug]   -enableRule:AppOffline
2023-08-14T08:48:51.6790093Z ##[debug]   -setParamFile=1692002930883_tempSetParameters.xml
2023-08-14T08:48:51.6790321Z ##[debug]   -setParamFile=1692002930883_tempSetParameters.xml
2023-08-14T08:48:51.6790539Z ##[debug]   -skip:Directory=App_Data
2023-08-14T08:48:51.6790754Z ##[debug]   -skip:Directory=App_Data
2023-08-14T08:48:51.6790958Z ##[debug]   -retryAttempts:6
2023-08-14T08:48:51.6791156Z ##[debug]   -retryAttempts:6
2023-08-14T08:48:51.6791368Z ##[debug]   -retryInterval:10000
2023-08-14T08:48:51.6791570Z ##[debug]   -retryInterval:10000
2023-08-14T08:48:51.6791797Z ##[debug]   -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
2023-08-14T08:48:51.6792060Z ##[debug]   -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
```

But right after the ToolRunner spawns the command, we see the output of the command containing the `""` wrapped around that parameter
```
2023-08-14T08:48:52.3065381Z [command]"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\a\5\{REDACTED}\{REDACTED}.zip' -dest:auto,ComputerName='https://{REDACTED}.scm.azurewebsites.net:443/msdeploy.axd?site={REDACTED}',UserName='${REDACTED}',Password='{REDACTED}',AuthType='Basic' "-setParam:name='IIS Web Application Name',value='{REDACTED}'" -enableRule:AppOffline -setParamFile=1692002930883_tempSetParameters.xml -skip:Directory=App_Data -retryAttempts:6 -retryInterval:10000 -userAgent:VSTS_21511a7c-2ec5-470e-8b8b-964cec1d8cd5_build_36_0
2023-08-14T08:48:52.8989800Z ##[debug]Exit code 4294967295 received from tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:52.8990715Z ##[debug]Exit code 4294967295 received from tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:52.8991402Z ##[debug]STDIO streams have closed for tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
2023-08-14T08:48:52.8991966Z ##[debug]STDIO streams have closed for tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
```

Backtracking a bit, we see that that in the AzureRmWebAppDeployment task that ToolRunner is called with `windowsVerbatimArguments: false` [here](https://github.com/microsoft/azure-pipelines-tasks-common-packages/blob/06c6234c71285be0fe9f3b544d1245aecaf5de47/common-npm-packages/webdeployment-common/deployusingmsdeploy.ts#L178C97-L178C128). In ToolRunner, we see that the command string is build via `this._getCommandString` [here](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/toolrunner.ts#L899C55-L899C55) and if we follow the path of what it will execute on `win32` platforms with `windowsVerbatimArguments` set to false, it will be this [here](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/toolrunner.ts#L178C43-L178C43). This will take us to `_windowsQuoteCmdArg` [here](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/toolrunner.ts#L395) where this line [here](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/toolrunner.ts#L414) checks if the arg itself needs quotes. If we check the [`cmdSpecialChars`](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/toolrunner.ts#L78C5-L78C5) it uses to define if it does encoding or not, we will see it will hit it for the argument in question, this encoding the argument in `""`. The logic for it is [here](https://github.com/microsoft/azure-pipelines-task-lib/blob/ff62c9b5a80b53e480e4dbfe87ace11244240dba/node/toolrunner.ts#L421C18-L421C18), from line 421 downwars and is per-argument encoded, which is what we see in the logs.

**Comment by BRGROSU** [comment]
To note, the IIS Web App Name is not actually called `{REDACTED}` as the `{` and `}` are included in those special characters in the array to decide encoding or not, the app name contains only `[0-9]`, `[a-z]` and `-` characters.

**Comment by leehadley** [comment]
Same problem here.

> Until fixed, we are mitigating by pegging the affected pipeline task to an unaffected version, e.g. `IISWebAppDeploymentOnMachineGroup@0.222.0` Instead of `IISWebAppDeploymentOnMachineGroup@0`

This got us working, thanks.

**Comment by timdiningrd** [comment]
Started failing for us also. No changes are our side, and it worked last week.

**Comment by timdiningrd** [comment]
Not sure how to get the workaround to work in Azure Pipelines YAML. Here is what my task is currently (that used to work):
![image](https://github.com/microsoft/azure-pipelines-tasks/assets/117851503/1a8465d1-9b44-43af-8487-490726f58187)

**Comment by wcheung-bhf** [comment]
change your task to

- task: AzureRmWebAppDeployment@4.225.1

**Comment by timdiningrd** [comment]
@wcheung-bhf That work! Thank you very much!

**Comment by DmitriiBobreshev** [comment]
The fix was published with Minor version 227 for AzureRmWebAppDeployment@4 and AzureRmWebAppDeployment@3.
@BRGROSU, @timdiningrd, @leehadley, could you please check, everything should work now?

**Comment by timdiningrd** [comment]
@DmitriiBobreshev Yes, it is working now (after I changed my task back to AzureRmWebAppDeployment@4).

Thank you!

**Comment by leehadley** [comment]
@DmitriiBobreshev Appears to be working for IISWebAppDeploymentOnMachineGroup@0 since update.

Thank you.

**Comment by BRGROSU** [comment]
Yes it seems version 4.227.0 fixed the deployments for us too. Thank you @DmitriiBobreshev.

**Comment by K-Cyganiak** [comment]
4.227.0 introduced new issue in our pipeline and no clear indication what cannot be found in error:

> "C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='D:\a\1\[redacted]\drop\Platform.Configuration.TDS_CM.wdp.zip' -dest:auto,ComputerName='redacted(redacted) -setParam:name='IIS Web Application Name',value='redacted' -skip:objectName=filePath,absolutePath=(?<!Views)(?<!Renderings)(?<!Layouts)\\Web\.config$ -retryAttempts:6 -retryInterval:10000 -enableRule:DoNotDeleteRule -userAgent:VSTS_3bfe60f7-585f-4e41-a693-18614d4838a3_build_704_0
> ##[debug]Exit code 1 received from tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
> ##[debug]Exit code 1 received from tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
> ##[debug]STDIO streams have closed for tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
> ##[debug]STDIO streams have closed for tool 'C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe'
> ##[debug]System.DefaultWorkingDirectory=D:\a\1\s
> ##[debug]Deployment Failed with Error: Error: The system cannot find the path specified.
> 
> ##[debug]task result: Failed
> ##[error]Error: The system cannot find the path specified.

Reverting to AzureRmWebAppDeployment@4.225.1 works fine.

**Comment by jockeEk** [comment]
I still get the broken `4.226.0` for some of my build, while some deploys get the working `4.225.1` .
The pipeline declares
```
AzureRmWebAppDeployment@4
```

**What controlls which minor and patch version that is choosen for the deploys?** I can of couse pin the version to `AzureRmWebAppDeployment@4.225.1` but I want to avoid that.

**Comment by kim3er** [comment]
Fixing the version to 4.225.1 has got us working, but this is obviously not ideal longterm.

**Comment by wcheung-bhf** [comment]
@jockeEk
Are you running old instances of your deployment pipeline or new deployments?
I had the same problem when retrying to run a failed stage but it worked when I re-ran the pipeline.

**Comment by jsayce** [comment]
> @jockeEk Are you running old instances of your deployment pipeline or new deployments? I had the same problem when retrying to run a failed stage but it worked when I re-ran the pipeline.

Thank you for this! I was repeatedly re-running the failed job, but it kept using 4.226.0 and getting the error. Starting a new pipeline run changed it to use 4.227.0, and the error has gone.

**Comment by jockeEk** [comment]
@DmitriiBobreshev no problem, bugs will happen. @wcheung-bhf first I did a re-run and then I started a new pipeline and still got the same behavior. And I still don't understand why.

Right now the version is pinned to `4.225.1` so I have not tried to go back to just `4` again.


My pipeline deploys the same application to a number of different app service plans that lives under different subscriptions. For all steps the step used `AzureRmWebAppDeployment@4` and for all deploys except one that ran `4.225.1`, on that specific deploy I got `4.226.0` (event after starting a new pipeline).

What decides the version when I only say `AzureRmWebAppDeployment@4`? And why can it be different for the same pipeline?

Thanks for all your quick replies this far, and for the contribution to this repo.
